File: languages/go/entity/key/access_control_type.go
package key

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

// AccessControlType represents the access control type of a key.
type AccessControlType int32

const (
	// ACCESS_CONTROL_NONE indicates that the key is not protected by access control.
	ACCESS_CONTROL_NONE AccessControlType = iota
	// ACCESS_CONTROL_TOTP indicates that the key is protected by a TOTP-based access control.
	ACCESS_CONTROL_TOTP AccessControlType = iota
	// ACCESS_CONTROL_SECRET indicates that the key is protected by a SECRET-based access control.
	ACCESS_CONTROL_SECRET AccessControlType = iota
)

var (
	AccessControlTypeFromProto = map[proto.AccessControlType]AccessControlType{
		proto.AccessControlType_NO_ACCESS_CONTROL: ACCESS_CONTROL_NONE,
		proto.AccessControlType_TOTP:              ACCESS_CONTROL_TOTP,
		proto.AccessControlType_SECRET:            ACCESS_CONTROL_SECRET,
	}

	AccessControlTypeToProto = map[AccessControlType]proto.AccessControlType{
		ACCESS_CONTROL_NONE:   proto.AccessControlType_NO_ACCESS_CONTROL,
		ACCESS_CONTROL_TOTP:   proto.AccessControlType_TOTP,
		ACCESS_CONTROL_SECRET: proto.AccessControlType_SECRET,
	}
)

File: languages/js/src/entity/key/access_control_type.ts
import * as keysEntitiesProto from "../../bridge/proto/bloock_keys_entities";

/**
 * Represents the access control type a cryptographic key.
 */
export enum AccessControlType {
  /**
   * Indicates that the key is not protected by access control.
   */
  NONE,
  /**
   * Indicates that the key is protected by a TOTP-based access control.
   */
  TOTP,
  /**
   * Indicates that the key is protected by a Secret-based access control.
   */
  SECRET
}

export namespace AccessControlType {
  export function toProto(
    type: AccessControlType
  ): keysEntitiesProto.AccessControlType {
    switch (type) {
      case AccessControlType.NONE:
        return keysEntitiesProto.AccessControlType.NO_ACCESS_CONTROL;
      case AccessControlType.TOTP:
        return keysEntitiesProto.AccessControlType.TOTP;
      case AccessControlType.SECRET:
        return keysEntitiesProto.AccessControlType.SECRET;
      default:
        return keysEntitiesProto.AccessControlType.NO_ACCESS_CONTROL;
    }
  }

  export function fromProto(
    type: keysEntitiesProto.AccessControlType | undefined
  ): AccessControlType {
    switch (type) {
      case keysEntitiesProto.AccessControlType.NO_ACCESS_CONTROL:
        return AccessControlType.NONE;
      case keysEntitiesProto.AccessControlType.TOTP:
        return AccessControlType.TOTP;
      case keysEntitiesProto.AccessControlType.SECRET:
        return AccessControlType.SECRET;
      default:
        return AccessControlType.NONE;
    }
  }
}
